stake-pool/single-pool/src/processor.rs

```diff
@@ -690,34 +712,33 @@ impl Processor {
             stake_history_info.clone(),
         )?;
 
-        let post_pool_stake = get_active_stake_state(pool_stake_info, clock.epoch)?
+        let (pool_stake_meta, pool_stake_state) = get_stake_state(pool_stake_info)?;
+        let post_pool_stake = pool_stake_state
             .delegation
-            .stake;
+            .stake
+            .saturating_sub(minimum_delegation);
         let post_pool_lamports = pool_stake_info.lamports();
-        msg!("Stake post merge {}", post_pool_stake);
-
-        // all lamports added, as a lamport difference, both stake and excess rent
-        let lamports_added = post_pool_lamports
-            .checked_sub(pre_pool_lamports)
-            .ok_or(SinglePoolError::ArithmeticOverflow)?;
+        msg!("Available stake post merge {}", post_pool_stake);
 
         // stake lamports added, as a stake difference
         let stake_added = post_pool_stake
             .checked_sub(pre_pool_stake)
             .ok_or(SinglePoolError::ArithmeticOverflow)?;
 
-        // excess rent, as all lamports less stake
-        let leftover_rent = lamports_added
-            .checked_sub(stake_added)
+        // we calculate absolute rather than relative to deposit amount to allow claiming lamports mistakenly transferred in
+        let excess_lamports = post_pool_lamports
+            .checked_sub(post_pool_stake)
+            .and_then(|amount| amount.checked_sub(pool_stake_meta.rent_exempt_reserve))
+            .and_then(|amount| amount.checked_sub(minimum_delegation))
             .ok_or(SinglePoolError::ArithmeticOverflow)?;
 
-        // sanity check: our calculated new stake matches the user prior stake
-        if stake_added != pre_user_stake {
+        // sanity check: we have not somehow gone below the minimum
+        if post_pool_stake < minimum_delegation {
             return Err(SinglePoolError::UnexpectedMathError.into());
         }
 
-        // sanity check: our calculated excess rent matches the user prior rent
-        if leftover_rent != user_unstaked_lamports {
+        // sanity check: our calculated new stake matches the user prior stake
+        if stake_added != pre_user_stake {
```

Comment:
omg

Reply:
but the `excess_lamports` calculation is still safe, right? because the user rent gets added to `pool_stake_state.delegation.stake`?

ah, but because of this we need to guard the withdraw call in case its zero

Reply:
wait but that also means calculating tokens off `stake_added` is wrong, and we need to use the difference between post and pre

Reply:
That's all correct